distributed/rpc/rl/distributed_model.py

```diff
@@ -0,0 +1,287 @@
+    """
+    return method(rref.local_value(), *args, **kwargs)
+
+
+def _remote_method(method, rref, *args, **kwargs):
+    r"""
+    a helper function to run method on the owner of rref and fetch back the
+    result using RPC
+    """
+    args = [method, rref] + list(args)
+    return rpc_sync(rref.owner(), _call_method, args=args, kwargs=kwargs)
+
+
+def _async_remote_method(method, rref, *args, **kwargs):
+    args = [method, rref] + list(args)
+    return rpc_async(rref.owner(), _call_method, args=args, kwargs=kwargs)
+
+class ObserverPolicy(nn.Module):
+    def __init__(self):
+        super(ObserverPolicy, self).__init__()
+        self.affine1 = nn.Linear(4, 128)
+        self.dropout = nn.Dropout(p=0.6)
+        self.affine2 = nn.Linear(128, 2)
+
+    def forward(self, x):
+        x = self.affine1(x)
+        x = self.dropout(x)
+        x = F.relu(x)
+        return self.affine2(x)
+
+
+class AgentPolicy(nn.Module):
+    def __init__(self):
+        super(AgentPolicy, self).__init__()
+
+        self.saved_log_probs = []
+        self.rewards = []
+
+    def forward(self, action_scores):
+        return F.softmax(action_scores, dim=1)
+
+
+class Observer:
+    def __init__(self):
+        self.id = rpc.get_worker_info().id
+        self.env = gym.make('CartPole-v1')
+        self.env.seed(args.seed)
+        self.policy = ObserverPolicy()
+
+    def get_gradients(self, ctx_id):
+        all_grads = dist_autograd.get_gradients(ctx_id)
+        out_grads = []
+        for p in self.policy.parameters():
+            out_grads.append(all_grads[p])
+        return out_grads
+
+    def update_model(self, params):
+        for p, new_p in zip(self.policy.parameters(), params):
+            with torch.no_grad():
+                p.set_(new_p)
+
+
+    def run_episode(self, agent_rref, n_steps):
+        r"""
+        Run one episode of n_steps.
+
+        Arguments:
+            agent_rref (RRef): an RRef referencing the agent object.
+            n_steps (int): number of steps in this episode
+        """
+        state, ep_reward = self.env.reset(), 0
+        for step in range(n_steps):
+            state = torch.from_numpy(state).float().unsqueeze(0)
+            action_scores = self.policy(state)
+            # send the state to the agent to get an action
+            action = _remote_method(Agent.select_action, agent_rref, self.id, action_scores)
+
+            # apply the action to the environment, and get the reward
+            state, reward, done, _ = self.env.step(action)
+
+            # report the reward to the agent for training purpose
+            _remote_method(Agent.report_reward, agent_rref, self.id, reward)
+
+            if done:
+                break
+
+class Agent:
+    def __init__(self, world_size):
+        self.ob_rrefs = []
+        self.agent_rref = RRef(self)
+        self.rewards = {}
+        self.saved_log_probs = {}
+        self.ob_policy = ObserverPolicy()
```

Comment:
The agent also creates a dummy ObserverPolicy so that it can use the same optimizer to update all model parameters. Note that this ObserverPolicy never participates in the forward or backward pass. It is only used to apply the summed gradients.